src/libs/ActiveClientManager/index.js

```diff
@@ -7,13 +7,29 @@ const clientID = Str.guid();
 const maxClients = 20;
 
 let activeClients;
+
+// Whether the current clientID is set
+let didInitialize = false;
+
+let resolveReadyPromise = null;
+
+// Keeps track of the ActiveClientManager's readiness in one place
+// so that multiple calls of isReady resolve the same promise
+const readyPromise = new Promise((resolve) => {
+    resolveReadyPromise = resolve;
+});
+
 Onyx.connect({
     key: ONYXKEYS.ACTIVE_CLIENTS,
     callback: (val) => {
         activeClients = !val ? [] : val;
         if (activeClients.length >= maxClients) {
             activeClients.shift();
-            Onyx.set(ONYXKEYS.ACTIVE_CLIENTS, activeClients);
+            Onyx.set(ONYXKEYS.ACTIVE_CLIENTS, activeClients).then(() => {
+                if (didInitialize) {
```

Comment:
I tried testing something out, and if I add `console.log(didInitialize)` right before this, it is never called with anything other than `true`. It feels to me like this variable isn't necessary. Is there some case where this is called and `didInitialize` would not be `true`?

Reply:
I also think this logic is in the wrong spot. Currently, this logic only runs once there are more than 20 clients. If you clear local storage, then you either have to refresh the page 20 times, or open 20 tabs, in order for this logic to be triggered.

Reply:
The last comment makes sense. I didn't realize it. wow.

Reply:
Seems like a good plan to me. I feel this is going to be little hactic. I belive that 
```
Onyx.connect({
    key: ONYXKEYS.ACTIVE_CLIENTS,
    callback: (val) => {
        activeClients = !val ? [] : val;
        if (activeClients.length >= maxClients) {
            activeClients.shift();
            Onyx.set(ONYXKEYS.ACTIVE_CLIENTS, activeClients).then(() => {
                isInitialized();
            });
        } else {
            isInitialized();
        }
    },
});
```
This is more accurate then as If the Max client condition is reached, then we can't be sure until the shift has happened.
```
function init() {
    Onyx.merge(ONYXKEYS.ACTIVE_CLIENTS, [clientID]).then(isInitialized);
}

Reply:
There is one more case when the client Leader change on tab-switch.

Reply:
It's probably technically more accurate, but I think it places the logic in a non-intuitive spot, and there isn't any issue with putting it in the `init()` method (the same thing is still achieved).

Reply:
Ok

Reply:
I tested it. First of it causes duplicate messages. Secondly, It hurt the code readability where we are using `ActiveClientManager.isClientLeader`. I agree with the point that your changes give the ActiveClientManger a clean look. But I think that this is going to be touched very less instead we will be using  `ActiveClientManager.isClientLeader` in a couple of places. and currently, we use it in a sync way. Making it a promise will break the sync flow.

Motive, why I added `isReady` callback to ActiveClientManager was that on page reload it was giving false results for client leadership and there was no client as a leader. We are not really concerned with the latter cases as we just want one client to process requests or send notifications for messages. it could be anyone. 

I have tried to change the current implementation thrice out of which I prefer Marc's suggestion of using `onReady` callback instead of an `isReady` promise.

Reply:
> and currently, we use it in a sync way. Making it a promise will break the sync flow.

Yeah, I get this. But that only means that we need to refactor the two spots that reference this method and they are both easily refactored. I don't think this should be a determination of what solution we should use.

I don't know if I've seen a full solution of Marc's suggestion. Is that what your branch is on now? I would like to make sure that I see it so I can be sure to have all the context.

Reply:
No, my Branch is not on that solution but here is a glimpse of that.
```
/**
 * Allow to run any task after ActiveClientManager has successfully initialized
 *
 * @returns {Promise<void>}
 */
function isReady() {
    return readyPromise;
}
```
TO
```
function onReady(callback) {
    resolveReadyCallback = callback;
}
```
Then at https://github.com/parasharrajat/Expensify.cash/blob/16c147f00de85fec995537dd4b5c4016c121e1ec/src/libs/ActiveClientManager/index.js#L30
call this callback. As we need to only call the callback at one place so this works.
```
 if (didInitialize) {
    resolveReadyPromise();
 }
 ```
 TO
 ```
  if (didInitialize) {
    resolveReadyCallback();
 }
```

Reply:
OK, thanks. I'm still really against `didInitialize` and `isReady` (or `onReady`). I'm not sure what it would take to convince me otherwise. I think both of those things make the code non-intuitive and we need to have a codebase that maintains low barriers to entry. If Joel (who is not traditionally a front-end engineer) can't understand what's going on, then it prevents him from lending a hand, and we are failing at being inclusive.

Reply:
OK. I will try to give one more attempt to what you suggested earlier and try to fix the duplication. But the `onReady` pattern is literally everywhere in our codebase.